### ISSUE TITLE
Icons are not rendering properly

### ISSUE THREAD

**Issue by athulvis** [diogocavilha/fancy-git#103]
Fancy git is a cool project. Anyway when I installed in in Debain Testing, the icons are not rendering properly

It looks as shown in the screenshot: 

![Screenshot_20220115_031316](https://user-images.githubusercontent.com/35709823/149589300-82932ca8-c8f5-4b16-be63-f81af3154153.png)

Font: `SauceCodePro NF 10pt`

Konsole version: `konsole 21.08.2`

Any help is appreciated. 

Thanks and regards,

Athul

**Comment by diogocavilha** [comment]
Hi, @athulvis 

Thank you for using Fancygit.

Are you facing that problem whatever the Fancygit font you use?

**Comment by athulvis** [comment]
@diogocavilha 

Yeah.. I tried with the different fonts mentioned in the documentation.

**Comment by tmarcokr** [comment]
Hi @athulvis , I'm facing the same problem. Have found a solution?

**Comment by athulvis** [comment]
@tmarcokr No.. its still same for me

**Comment by ljonya** [comment]
I have the same problem not only with konsole, but also with vscode. Everything works fine in Kitty terminal

**Comment by ljonya** [comment]
Hi @athulvis, @tmarcokr do you have this problem with vscode terminal on linux?

**Comment by diogocavilha** [comment]
Hi guys. Hope you all are doing good. 

This is an issue when using Fancygit on KDE, unfortunately.
I've been looking for the solution but I can't solve it yet.
I'm so sorry for that :(

**Comment by diogocavilha** [comment]
Hi guys. Hope you're all doing good.

This is an issue when using Fancygit on KDE, unfortunately.
I've been looking for the solution but I can't solve it yet.
I'm so sorry for that :(

**Comment by diogocavilha** [comment]
Maybe a possible workaround would be override the default values by using: [https://github.com/diogocavilha/fancy-git#gear-advanced-tweaking---override-some-default-config-values](https://github.com/diogocavilha/fancy-git#gear-advanced-tweaking---override-some-default-config-values)

**Comment by sbsud** [comment]
Same issue with mac terminal and iterm (and vscode) I was able to override defaults and get most of the icons to load. But would have been good for it to work without the overrides. The first image still does not load and your document does not specify the override variable for it. Also very cumbersome to find the icon from the font and update the environment variables.

![image](https://user-images.githubusercontent.com/2920543/158775039-5bae422b-efcc-4964-8443-e0e8516cc827.png)

![image](https://user-images.githubusercontent.com/2920543/158775429-01185615-ac1c-4c97-9b03-5025d6be54d5.png)

**Comment by yasindagasan** [comment]
same issue in wsl

**Comment by EDIflyer** [comment]
Same issue in Windows Terminal - have tried other nerdfonts packages too -
![image](https://user-images.githubusercontent.com/13610277/174502919-1f0f58a9-b0bf-4861-9aea-81652648af7c.png)

Any solutions/suggestions welcome!

**Comment by EDIflyer** [comment]
OK in the end found `Sauce-Code-Pro-Nerd-Font-Complete-Windows-Compatible.ttf` seems to be the only one that works in Windows Terminal - tried all the others!  Icons now working but I'm finding it's quite a 'light' font...
![image](https://user-images.githubusercontent.com/13610277/174688677-4d918732-c246-4f40-8ae8-63738aca1318.png)

There are some rather nice ones on Nerd Font with PowerLine glyphs (e.g Caskaydia Cove Nerd Font from https://github.com/ryanoasis/nerd-fonts/releases/download/v2.1.0/CascadiaCode.zip) - is there any way of supporting these options @diogocavilha? Would be great to have a broader range of fonts to pick from :)

**Comment by EDIflyer** [comment]
Thanks @hongsoog - however I'm using Windows Terminal for an SSH connection to a virtual server, it looks like your suggestion is for running on a local machine (apologies if misunderstood).

**Comment by diogocavilha** [comment]
> The issue is related to the terminal emulator capability to render the emoji in the fancygit recommended fonts.
> 
> I have tries several terminal enumlaters, the find '**lxterminal**' shows the icons without any problem.
> 
> TL; DR
> 
> ```shell
> $ sudo apt install lxterminal
> $ lxterm&
> ```
> 
> Then go to **Edit - Preferences** menu of lxterminal window, select **Style** tab and choose your favorate fonts(e.g. SauceCodePro NF Regulat) and click OK. Done!!
> 
> * Tested on Unbuntu 20.04 LTS with remote connection with MobaXterm from Windows 10.
> 
> Cheers!!

@hongsoog thank you so much for that solution!

**Comment by EDIflyer** [comment]
OK, thanks @diogocavilha - I can totally appreciate the difficulties in supporting multiple fonts I'll perhaps try something different then as for normal font usage I much prefer the Caskaydia Cove and similar - the ones that work don't render quite a nicely for text for me.  As mentioned @hongsoog's suggestion only works if running on a local Linux machine, whereas as I'm SSHing in from Windows Terminal. Cheers :)

**Comment by harnen** [comment]
The same issue on the default Ubuntu 22.04 terminal. I tried all the suggested fonts but not a single one displays all the icons.

**Comment by harnen** [comment]
> > The same issue on the default Ubuntu 22.04 terminal. I tried all the suggested fonts but not a single one displays all the icons.
> 
> Hi, @harnen.
> 
> Nor even the SourceCodePro font worked?

Ha, I've just reinstalled the SourceCodePro font and set it system-wide (it doesn't show in the terminal)  and it worked. I might have done something incorrectly before... Anyway, thanks a lot!

**Comment by diogocavilha** [comment]
> > > The same issue on the default Ubuntu 22.04 terminal. I tried all the suggested fonts but not a single one displays all the icons.
> > 
> > 
> > Hi, @harnen.
> > Nor even the SourceCodePro font worked?
> 
> Ha, I've just reinstalled the SourceCodePro font and set it system-wide (it doesn't show in the terminal) and it worked. I might have done something incorrectly before... Anyway, thanks a lot!

Oh, that's great! :smiley:

**Comment by pwerden** [comment]
I faced the same issue with icons not showing correctly. (Debian 11, bash - KDE terminal Version 20.12.3) 
For me the only solution that worked was to use Font "SourceCodePro+Powerline+Awesome"

**Comment by Nootaku** [comment]
Hello,

Could you please give us access to the original icons that you are using ?<br/>I have included the missing icons in `theme-functions.sh` by copying unicode icons from the web. While this works, I cannot find the git-branch icons.

Also as an improvement idea, maybe you could regroup all icons in one file and import them where necessary.

**Comment by hajtuJ** [comment]
For wsl users:
> Hi @bsides I've used several fonts from [www.nerdfonts.com](http://www.nerdfonts.com): SourceCodePro.zip, DroidSansMono.zip, FiraMono.zip, Noto.zip. ..and they didn't work.
> 
> Now, I've changed my workflow, now I'm using Ubuntu 20.04 laptop with the standard Terminal and there I tried above fonts without success and the only fonts have worked were the included in FancyGit repo (https://github.com/diogocavilha/fancy-git/raw/master/fonts/SourceCodePro%2BPowerline%2BAwesome%2BRegular.ttf).
> 
> Regards.
Ref: https://github.com/diogocavilha/fancy-git/issues/70#issuecomment-671361925

**Comment by datenfalke** [comment]
> I faced the same issue with icons not showing correctly (Debian 11, bash - KDE terminal Version 20.12.3). For me the only solution that worked was to use font "SourceCodePro+Powerline+Awesome+Regular"

this!

**Comment by k-willowhawk** [comment]
Had the same issue running in iTerm2 on macOS ventura.  Same solution, use the linked SourceCodePro-Powerline-Awesome-Regular font mentioned in previous comments.

**Comment by nihklas** [comment]
For anyone still having this Problem on MacOS:
I had the Issue, that whichever font i used, some icons were missing. I preferred the Jetbrains Mono Font, but it missed all the status icons. When using the SourceCodePro Font i had every status icon, but any Separator Icon except the standard one was missing from there. 
I fixed it by patching a new Jetbrains Mono Font myself:

1. I installed FontForge
```bash
brew install --cask fontforge
```

2. Then i opened the Jetbrains Mono Font shipped with fancy-git, as well as SourceCodePro in separate windows.
3. Then i looked up all missing or wrong Unicode Code Points:
```bash
printf '%X\n' "'<copy-paste-icon-here>"
```
(Note the `'` before the icon)

4. for each missing icon, i looked through the SourceCodePro font and copy pasted it into the Jetbrains Font
5. When i was ready, i pressed `CTRL-SHIFT-G` and generated the font under a new name
    - Make sure you select TrueType as the format tho
6. I opened the font in Font Book and pressed install
7. Lastly, i selected the font in my iTerm2 Profile

**Comment by nihklas** [comment]
I opened a Pull-Request to add the missing icons into all Font-files (see #152)

**Comment by maheshkn400** [comment]
> Same issue with mac terminal and iterm (and vscode). I was able to override defaults and get most of the icons to load. But would have been good for it to work without the overrides. The first icon still does not load and your documentation does not specify the override variable for it. Also very cumbersome to find the icon from the font and update the environment variables. Can you comment on why the icons may not be loading? I am using the Jetbrains mono nerd medium font.
> 
> ![image](https://user-images.githubusercontent.com/2920543/158775039-5bae422b-efcc-4964-8443-e0e8516cc827.png)
> 
> ![image](https://user-images.githubusercontent.com/2920543/158775429-01185615-ac1c-4c97-9b03-5025d6be54d5.png)

Yes, it is. Add the following to your bashrc or profile based on your Linux version, then restart the console and navigate to any git repository.

export FANCYGIT_ICON_GIT_REPO=""
export FANCYGIT_ICON_LOCAL_BRANCH=""
export FANCYGIT_ICON_LOCAL_REMOTE_BRANCH="⇄"
export FANCYGIT_ICON_MERGED_BRANCH="✔"
export FANCYGIT_ICON_HAS_STASHES="📦"
export FANCYGIT_ICON_HAS_UNTRACKED_FILES="?"
export FANCYGIT_ICON_HAS_CHANGED_FILES="Δ"
export FANCYGIT_ICON_HAS_ADDED_FILES="+"
export FANCYGIT_ICON_HAS_UNPUSHED_COMMITS="↑"
export FANCYGIT_ICON_VENV="🐍"

**Comment by greigs** [comment]
In case anyone is using WSL and the solution from @maheshkn400 doesn't work, try the environment variables in [this gist](https://gist.github.com/greigs/665cd0f2323b1750941b08cfff688d1c).
